File: Searching/binarySearch.py
def binarySearch(array, target, ascending_order = True):
	"""
	Search a sorted list 'array' for 'target'. 
	Returns index of 'target' if found. Else -1 to indicate not found.

	Parameters:
		array (list): sorted list to search.
		target (number): the element to search for.
		ascending_order (bool, optional): how list is sorted.

	Returns:
		integer: index of occurrence; or -1 if not found.
	
	Behavior:
		Search a sorted list for the specified target.

	Usage:
		/ Ascending ordered list
		binarySearch(array, target)

		/ Descending ordered list
		binarySearch(array, target, ascending_order = False)
	"""
	low = 0
	high = len(array) - 1
	middle = (high - low) // 2

	while low <= high:
		middle = int((low + high) / 2)

		# List in ascending order:
		if ascending_order:
			if array[middle] < target:
				low = middle + 1
			elif array[middle] > target:
				high = middle - 1
			else:
				return middle

		# List in descending order:
		else:
			if array[middle] > target:
				low = middle + 1
			elif array[middle] < target:
				high = middle - 1
			else:
				return middle

	# Target is not in the list/array
	return -1

File: Searching/binarySearch.c
#include <stdio.h>

int binarySearch(int array[], int size, int target);

int main(void)
{
    // Make an array to search; and a target to find.
    int array_[] = {2, 7, 8, 12, 15, 22, 25, 28};
    int size = 8;
    int target = 12;

    // Store index/-1 in variable 'result'; and print message.
    int result = binarySearch(array_, size, target);
    if (result != -1)
        printf("Found at index # %i\n", result);
    else
        printf("Did not found!\n");

    return 0;
}

int binarySearch(int array[], int size, int target)
{
    // Initialize variables.
    int low = 0;
    int high = size - 1;
    int middle;

    while (low <= high)
    {
        // Calculate next middle.
        middle = (low + high) / 2;

        if (array[middle] < target)
            low = middle + 1;
        else if (array[middle] > target)
            high = middle - 1;
        else
            return middle;
    }

    // Target is NOT in the array.
    return -1;
}
